editor/bitmap.ts

```diff
@@ -0,0 +1,95 @@
+namespace mkcd {
+    export type Coord = [number, number];
+
+    /**
+     * 16-color sprite
+     */
+    export class Bitmap {
+        protected buf: Uint8Array;
+
+        constructor(public width: number, public height: number, public x0 = 0, public y0 = 0) {
+            this.buf = new Uint8Array(Math.ceil(width * height / 2));
+        }
+
+        set(col: number, row: number, value: number) {
+            if (col < this.width && row < this.height && col >= 0 && row  >= 0) {
+                const index = this.coordToIndex(col, row);
+                this.setCore(index, value);
+            }
+        }
+
+        get(col: number, row: number) {
+            if (col < this.width && row < this.height && col >= 0 && row  >= 0) {
+                const index = this.coordToIndex(col, row);
+                return this.getCore(index);
+            }
+            return 0;
+        }
+
+        copy(col = 0, row = 0, width = this.width, height = this.height): Bitmap {
+            const sub = new Bitmap(width, height);
+            sub.x0 = col;
+            sub.y0 = row;
+            for (let c = 0; c < width; c++) {
+                for (let r = 0; r < height; r++) {
+                    sub.set(c, r, this.get(col + c, row + r));
+                }
+            }
+            return sub;
+        }
+
+        apply(change: Bitmap) {
+            for (let c = 0; c < change.width; c++) {
+                for (let r = 0; r < change.height; r++) {
+                    this.set(change.x0 + c, change.y0 + r, change.get(c, r));
+                }
+            }
+        }
+
+        protected coordToIndex(col: number, row: number) {
+            return col + row * this.width;
+        }
+
+        protected getCore(index: number) {
+            const cell = Math.floor(index / 2);
+            if (index % 2 === 0) {
+                return this.buf[cell] & 0xf;
+            }
+            else {
+                return (this.buf[cell] & 0xf0) >> 4;
+            }
+        }
+
+        protected setCore(index: number, value: number) {
+            const cell = Math.floor(index / 2);
+            if (index % 2 === 0) {
+                this.buf[cell] = (this.buf[cell] & 0xf0) | (value & 0xf);
+            }
+            else {
+                this.buf[cell] = (this.buf[cell] & 0x0f) | ((value & 0xf) << 4);
+            }
+        }
+    }
+
+    export class Bitmask {
+        protected mask: Uint8Array;
+
+        constructor(public width: number, public height: number) {
+            this.mask = new Uint8Array(Math.ceil(width * height / 8));    
+        }
+
+        set(col: number, row: number) {
+            const cellIndex = col + this.width * row;
+            const index = Math.floor(cellIndex / 8);
+            const offset = cellIndex % 8;
```

Comment:
Not like it matters, but `>>3` and `&7` would be probably faster, since mod in JavaScript can be floating point.